### PR TITLE
CROS-125: add block buffer

### PR DESCRIPTION
**Pull request by RuilaiZhang**

### CODE REVIEW

File: src/sync-pending-events.ts [hunk 1]
```diff
@@ -132,7 +135,7 @@ export class EventTracker {
     let latestBlockNumber = await getLatestBlockNumber(this.source);
     let trueLatestBlockNumber = await getValidBlockNumber(this.source, lastCheckedBlockNumber, latestBlockNumber);
     if (trueLatestBlockNumber) {
-      latestBlockNumber = trueLatestBlockNumber;
+      latestBlockNumber = trueLatestBlockNumber - this.bufferBlocks;
```
Comment (yymone) [inline src/sync-pending-events.ts]: Have you consider if trueLatestBlockNumber  is less than 20 scenario?

File: src/sync-pending-events.ts [hunk 2]
```diff
@@ -132,7 +135,7 @@ export class EventTracker {
...
     }else {
       latestBlockNumber = lastCheckedBlockNumber;
     }
```
Comment (yymone) [inline src/sync-pending-events.ts]: Here, as we had added a new logic to "- this.bufferBlocks",
So this piece of codes from 135 to 141 need to be a function (maybe the logic of the parameter become fromBlockNumber? Leave to you to decide the name) and you can add unit test for this function.
In the unit tests you need cover all edge cases like trueLatestBlockNumber is less than bufferBlocks

File: src/sync-pending-events.ts [hunk 3]
```diff
@@ -129,17 +132,13 @@ export class EventTracker {
...
-    }else {
-      latestBlockNumber = lastCheckedBlockNumber;
-    }
     if(lastCheckedBlockNumber < this.eventSyncStartingPoint) {
       lastCheckedBlockNumber = this.eventSyncStartingPoint;
       console.log("Chain Height Before Deployment: " + lastCheckedBlockNumber + " Is Used");
     }
+    let latestBlockNumber = await getLatestBlockNumberFromSource(this.source);
+    let trueLatestBlockNumber = await getValidBlockNumber(this.source, lastCheckedBlockNumber, latestBlockNumber);
+    latestBlockNumber = await getReliableBlockNumber(trueLatestBlockNumber, lastCheckedBlockNumber, this.bufferBlocks);
```
Comment (yymone) [inline src/sync-pending-events.ts]: I think change the param name to reliableBlockNumber will be better

File: src/sync-pending-events.ts [hunk 4]
```diff
@@ -129,20 +132,16 @@ export class EventTracker {
...
-      latestBlockNumber = trueLatestBlockNumber;
-    }else {
-      latestBlockNumber = lastCheckedBlockNumber;
-    }
     if(lastCheckedBlockNumber < this.eventSyncStartingPoint) {
       lastCheckedBlockNumber = this.eventSyncStartingPoint;
       console.log("Chain Height Before Deployment: " + lastCheckedBlockNumber + " Is Used");
     }
+    let latestBlockNumber = await getLatestBlockNumberFromSource(this.source);
+    let trueLatestBlockNumber = await getValidBlockNumber(this.source, lastCheckedBlockNumber, latestBlockNumber);
```
Comment (xgaozoyoe) [inline src/sync-pending-events.ts]: right is getValidBlockNumber(), and left is trueLatestBlockNumber. Better use the same name convention, either "valid" or "trueLatest", otherwise it is very confusing what is the meaning of getValidBlockNumber. (In this case, I guess the name of getValidBlockNumber is misleading)

File: src/sync-pending-events.ts [hunk 5]
```diff
@@ -254,7 +255,15 @@ export const getweb3 = {
   }
 }
 
-async function getLatestBlockNumber(provider: string) {
+export async function getReliableBlockNumber(trueLatestBlockNumber: any, lastCheckedBlockNumber: number, bufferBlocks: number) {
```
Comment (xgaozoyoe) [inline src/sync-pending-events.ts:260]: do we really need to export this (only for test?）

Comment (xgaozoyoe) [reply src/sync-pending-events.ts:260]: the same problem exists for getValidBlockNumber.

Comment (yymone) [reply src/sync-pending-events.ts:260]: It is for unit test and looks like no better simple solution so just leave it

File: src/sync-pending-events.ts (hunk 4 above)
Comment (yymone) [reply src/sync-pending-events.ts]: Need change and make it consistent

File: src/client.ts [hunk 6]
```diff
@@ -53,6 +53,10 @@ export class DelphinusContract {
     let pastEvents = [];
     let start = fromBlock;
     let end = 0;
+    if(fromBlock > toBlock){
+      console.log("No New Blocks Found From:"+ fromBlock);
+      return {"events": [], "breakpoint": fromBlock}
```
Comment (yymone) [inline src/client.ts]: Is this correct?
Assume it is just normal scenario and latest block number not increased.
For example, lastCheckedBlock = 10, and reliableBlockNumber=10
So we didn't check 11,
However by this code, it will return 11 and await db.updatelastCheckedBlockNumber(pastEvents.breakpoint); will update last checked block to 11.

Comment (yymone) [reply src/client.ts]: I think adjust the breakpoint value is not a good solution and easy to have bug.
It should do not call API and return null.

And at the place which call it, add code to treat null.
Never use any logic like -1 as it is not maintainable and scalable. (Think about if in the future we change the from from +1 to other number, it is very hard to figure out and fix the code logic.

**Review by yymone — APPROVED** [review]
